helix-view/src/document.rs

```diff
@@ -163,23 +172,107 @@ impl Document {
         }
     }
 
+    #[inline(always)]
+    fn calculate_decode_capacity(decoder: &mut encoding_rs::Decoder, buf: &[u8]) -> usize {
+        // Will never panic because the buffer size will never be close to usize::MAX in the
+        // context it is used.
+        let valid_up_to = encoding_rs::Encoding::utf8_valid_up_to(buf);
+        let max_buffer_len = decoder
+            .max_utf8_buffer_length(buf.len() - valid_up_to)
+            .unwrap();
+        valid_up_to + max_buffer_len
+    }
+
+    #[inline(always)]
+    fn calculate_encode_capacity(encoder: &mut encoding_rs::Encoder, buf: &[u8]) -> usize {
+        // Will never panic because the buffer size will never be close to usize::MAX in the
+        // context it is used.
+        let valid_up_to = encoding_rs::Encoding::utf8_valid_up_to(buf);
+        let max_buffer_len = encoder
+            .max_buffer_length_from_utf8_if_no_unmappables(buf.len() - valid_up_to)
+            .unwrap();
+        (valid_up_to + max_buffer_len).next_power_of_two()
+    }
+
     // TODO: async fn?
-    pub fn load(path: PathBuf) -> Result<Self, Error> {
-        use std::{fs::File, io::BufReader};
+    /// Open a file from a path, or create a new file if it doesn't exist.
+    pub fn open(path: PathBuf) -> Result<Self, Error> {
+        use std::{
+            fs::File,
+            io::{BufReader, Read},
+        };
 
-        let doc = if !path.exists() {
-            Rope::from("\n")
-        } else {
-            let file = File::open(&path).context(format!("unable to open {:?}", path))?;
-            let mut doc = Rope::from_reader(BufReader::new(file))?;
-            // add missing newline at the end of file
-            if doc.len_bytes() == 0 || doc.byte(doc.len_bytes() - 1) != b'\n' {
-                doc.insert_char(doc.len_chars(), '\n');
-            }
-            doc
+        if !path.exists() {
+            return Ok(Self::new());
+        }
+
+        // Create a 64kb buffer for reading the file into.
+        let mut buf = [0u8; 1 << 16];
+        let mut builder = RopeBuilder::new();
+        let mut file = BufReader::new(File::open(&path)?);
+        let (encoding, mut decoder, is_last) = {
+            let read = file.read(&mut buf)?;
+            let is_last = read < buf.len();
+            let mut encoding_detector = chardetng::EncodingDetector::new();
+            encoding_detector.feed(&buf, is_last);
+
+            let encoding = encoding_detector.guess(None, true);
+            let decoder = encoding.new_decoder();
+            builder.append(&encoding.decode(&buf[..read]).0);
```

Comment:
I think it would be clearer to move all of the actual decoding into the loop below.  Then any special cases that might come up can be handled in one place.